### DISCUSSION TITLE
utorrent-block-xunlei-v2.4

### DISCUSSION THREAD

**Discussion by ShenHongFei** [ShenHongFei/utorrent-block-xunlei#26] [General]
## 使用方法
1. 确保 uTorrent 是 3.0 以上版本（不支持 2.x 版本），并开启 WebUI (网页界面)  
1.1 打开 uTorrent 设置 > 高级 > 网页界面  
1.2 选上 "启用网页界面", 并在下方填写用户和密码, 记下来, 作为命令行 `--username` 和 `--password` 参数的值  
2. 在 "uTorrent 设置 > 连接 > 监听端口 > 传入连接所使用的端口" 中查看设置的端口号，记下来，作为命令行 `--port` 参数的值
3. 在 uTorrent 数据目录中（路径可能是 `C:/Users/<用户名>/AppData/Roaming/uTorrent/ipfilter.dat` ，该目录下应该有 resume.dat 文件存在）保证 ipfilter.dat 文件存在（若不存在则新建空白 ipfilter.dat 文件），脚本会在原有 ipfilter.dat 文件内容之后添加被屏蔽的 IP，不影响已有内容及其功能
4. 在 releases 中下载 `utorrent-block-xunlei.exe`, 下载地址: [https://github.com/ShenHongFei/utorrent-block-xunlei/releases](https://github.com/ShenHongFei/utorrent-block-xunlei/releases)
5. 查看用法: `utorrent-block-xunlei.exe --help`
```text
Usage: utorrent-block-xunlei --port 50050 --username tom --password 123456 --ipfilter "C:/Users/tom/AppData/Roaming/uTorrent/ipfilter.dat"

Xunlei Blocker of uTorrent

Options:
  -V, --version                output the version number
  --hostname <hostname>        可选参数, uTorrent WebUI 主机的 hostname, 默认为本机 127.0.0.1, 也可设置远程主机 IP 或域名 (default: "127.0.0.1")
  --port <port>                必传参数, "uTorrent 设置 > 连接 > 监听端口 > 传入连接所使用的端口" 中设置的端口号
  --username <username>        必传参数, "uTorrent 设置 > 高级 > 网页界面 > 身份验证 > 用户" 中设置的用户名
  --password <password>        必传参数, "uTorrent 设置 > 高级 > 网页界面 > 身份验证 > 密码" 中设置的密码
  --ipfilter <ipfilter>        必传参数, uTorrent 数据目录中 ipfilter.dat 文件的完整路径，如: C:/Users/tom/AppData/Roaming/uTorrent/ipfilter.dat
  --interval <interval>        可选参数, 检测 peers 的间隔（秒）, 默认每隔 20 秒检测并屏蔽一次 (default: "20")
  --interval-reset <interval>  可选参数, 间隔 interval 秒自动重置当前时间间隔内被动态屏蔽的 IP，默认间隔 2 小时 (default: "7200")
  -h, --help                   display help for command
```

6. 启动 blocker (修改下参数的值): `utorrent-block-xunlei --port 50050 --username shf --password 123456 --ipfilter "C:/Users/shf/AppData/Roaming/uTorrent/ipfilter.dat"`

7. (可选，不影响屏蔽功能) 在 uTorrent 中开启屏蔽日志，查看被屏蔽的连接请求
    4.1 在 uTorrent 下方的日志面板中点击右键
    4.2 勾选 `记录用户通讯信息 > 记录拦截连接`
    4.3 有被屏蔽的用户连接时可以在 uTorrent 日志面板中看到类似下面的日志
```text
[2018-11-22 19:03:43]  Loaded ipfilter.dat (51 entries)
[2018-11-22 19:03:46]  IpFilter blocked peer 223.81.192.235
[2018-11-22 19:03:49]  IpFilter blocked peer 223.81.192.235
[2018-11-22 19:04:06]  IpFilter blocked peer 223.81.192.235
[2018-11-22 19:04:21]  IpFilter blocked peer 183.25.54.216
[2018-11-22 19:04:46]  IpFilter blocked peer 223.81.192.235
...
```

8. 实时控制，查看状态，REPL
在 `utorrent-block-xunlei` 命令行窗口中输入 `utorrent` 可通过 REPL 查看其属性，调用其方法
如输入 `utorrent.print_blockeds()` 可查看被动态屏蔽的 IP

9. 退出 utorrent-block-xunlei 时需在命令行窗口中输入 `exit()`, 以保证 `ipfilter.dat` 恢复到运行前的状态，不残留动态屏蔽的 IP

<hr /><em>This discussion was created from the release <a href='https://github.com/ShenHongFei/utorrent-block-xunlei/releases/tag/v2.4'>utorrent-block-xunlei-v2.4</a>.</em>

**Comment by tonyhsie** [comment]
> @SeaHOH @tonyhsie v2.4 已修复，再试试吧

先感謝您的辛勤勞動

我試了一下v2.4，ipfilter.dat 不會被清為0了
不過有三個小問題

1. 如何正常退出? 輸入 .exit 之後不會退出，一樣還是要按 Ctrl-C 幾次才會退出，但ipfilter.dat新增的內容仍舊還在，並沒有因為退出而清除
2. PC speaker 有機率會每20秒發出一下嗶聲，就跟 utorrent 出錯崩潰時所發出的聲音一樣，但實際上 utorrent 仍然運行地好好的，blocker本身也沒有錯誤訊息，不曉得是什麼問題
3. 現在v2.4版運行的邏輯跟之前1.0版的邏輯似乎不太一樣，我不太明白現在ipfilter.dat跟blocker跟ut三者間是怎麼互動的。據我的理解，1.0版是每30秒偵測到xunlei的IP之後，會把這些IP馬上寫入ipfilter.dat，然後讓ut重新載入ipfilter，以達成block的目的(有錯也請指正)

但現在2.X版在偵測到xunlei IP之後，好像也不一定會寫入ipfilter.dat?

以圖來說明，下圖中間可見 blocker 偵測到一整排應該要封鎖的IP
但背後是ipfilter.dat的內容，左下角卻只有一個 IP，這樣的 behavior 我個人不知道該怎麼解讀，有點混亂

![未命名](https://user-images.githubusercontent.com/52758827/129440934-8ce6e4ad-415f-4c55-ae31-27e4441885bc.png)

**Reply by tonyhsie** [reply]
這裡也一併附上 v1.0 版的 blocker 執行結果/ipfilter.dat 內容圖
方便對照兩者差異


![未命名2](https://user-images.githubusercontent.com/52758827/129440756-52ad1e3f-8180-4b70-889c-32f5d10cf57c.png)

**Reply by ShenHongFei** [reply]
> 不過有三個小問題
> 
> 1. 如何正常退出? 輸入 .exit 之後不會退出，一樣還是要按 Ctrl-C 幾次才會退出，但 ipfilter.dat 新增的內容仍舊還在，並沒有因為退出而清除
> 2. PC speaker 有機率會每 20 秒發出一下嗶聲，就跟 utorrent 出錯崩潰時所發出的聲音一樣，但實際上 utorrent 仍然運行地好好的，blocker 本身也沒有錯誤訊息，不曉得是什麼問題
> 3. 現在 v2.4 版運行的邏輯跟之前 1.0 版的邏輯似乎不太一樣，我不太明白現在 ipfilter.dat 跟 blocker 跟 ut 三者間是怎麼互動的。據我的理解，1.0 版是每 30 秒偵測到 xunlei 的 IP 之後，會把這些 IP 馬上寫入 ipfilter.dat，然後讓 ut 重新載入 ipfilter，以達成 block 的目的 (有錯也請指正)

@tonyhsie 
问题 1. 在窗口里直接输入 `exit()` 正常退出， 不能 `.exit`, 也不能按 ctrl + c
问题 2. “PC speaker 有機率會每 20 秒發出一下嗶聲”这个问题我找个虚拟机研究下，等会回复你
问题 3. 这个是符合预期的，新版修改了一下屏蔽方法，如下

### 屏蔽策略
- 下方屏蔽列表中的非法客户端才会被屏蔽
- 任务处于做种状态时：屏蔽吸血速度大于 `10 KB/s` 的非法客户端，将带宽留给合法客户端
- 任务处于下载状态时，对有上传流量的非法客户端放宽处理：只屏蔽累计吸血量大于 `10 × 累计上传量 + 5MB` 的非法客户端

理由:  
过于频繁的重载 ipfilter.dat 可能影响上传、下载速度，因此设置了一个阈值，忽略已连接但吸血速度过低的非法客户端
下载时部分非法客户端也有积极上传，为了提升我方的下载速度，暂时赦免他们，也给他们上传


### 屏蔽列表
-XL0012-***  
Xunlei/***  
7.x.x.x
Xfplay  
FDM  
dandanplay  
Mozilla  
go.torrent
aria2

**Reply by ShenHongFei** [reply]
第二个问题我看了下，下个版本修复吧，应该不会经常出现。

**Reply by SeaHOH** [reply]
过滤下客户端名称，有时候会包含 peerID，那个声音估计是打印了`\x07`。

**Reply by ShenHongFei** [reply]
应该是的

**Reply by SeaHOH** [reply]
有部分迅雷完全不会上传，特征明显，要尽早判断不要推迟屏蔽。
20秒检测间隔可能偏高，现在家庭宽带都非常快了，20秒足以吸几十MB。

**Reply by tonyhsie** [reply]
另外建議一下

2.x 版的訊息量太多
雖然實際上一個用戶只印一行信息，但因為過長，顯示在屏幕上變成三行
就算開了全螢幕，也要佔兩行

這樣太過眼花繚亂，失去了打印信息以便閱讀的本意
最好還是一個用戶僅打印出一行以內的信息

PS. 能打印出地名，單位信息，還真讓人耳目一新
不過其它速度/上下載量/種子名稱，這些信息就太過細節，如果有必要打印，也請儘量縮簡篇幅

**Reply by ShenHongFei** [reply]
你可以考虑换个等宽字体，然后把字号调小点，推荐用 ConEmu，理想效果是下面这样的
![](https://github.com/ShenHongFei/utorrent-block-xunlei/raw/master/preview/monitor-2.png)

**Reply by ShenHongFei** [reply]
或者你也可以试试 Windows Terminal

**Reply by tonyhsie** [reply]
> 你可以考虑换个等宽字体，然后把字号调小点，推荐用 ConEmu，理想效果是下面这样的
> ![](https://github.com/ShenHongFei/utorrent-block-xunlei/raw/master/preview/monitor-2.png)

CMD 只能用等寬字體沒錯，字號調小也是難以閱讀，並不理想

不過你這是 21:9 的螢幕嗎?

**Reply by ShenHongFei** [reply]
就普通的 16:9 屏幕，用 ConEmu，字号调小了点

**Reply by tonyhsie** [reply]
> 就普通的 16:9 屏幕，用 ConEmu，字号调小了点

可惜我不是用 4K 螢幕啊，不是字號調小的問題了

**Comment by Teddy2839** [comment]
你好，我在使用时出现了一个问题，在第4~5步的地方，下载了utorrent-block-xunlei.exe程序，双击或者管理员方式打开后，程序闪了不到一秒钟就消失了，我使用win7系统，不知道是系统问题还是需要安装额外的软件？谢谢

**Comment by samlee111** [comment]
就挺尴尬。。最近重装了系统，想着软件都翻新一下，发现这个新版的不如20年的旧版好用？旧版的代码就一个js文件，我自己调整也很方便，现在单个exe文件说实话我不会搞。。
遇到的问题有：
·启动后有概率不工作，不写入屏蔽文件
·用户列表里有XL0012(ipv6)却没有被屏蔽，但有时又能屏蔽
·µTorrent自己也被屏蔽了
·写入一次屏蔽后程序死机不再执行
![img-Text](https://s1.ax1x.com/2022/09/02/vI5Eff.png)

旧版运行良好
![img-Text](https://s1.ax1x.com/2022/09/02/vI5AtP.png)

系统环境：
版本	Windows 10 专业版
版本号	21H2
安装日期	‎2022/‎8/‎22
操作系统内部版本	19044.1889
体验	Windows Feature Experience Pack 120.2212.4180.0
µTorrent 3.5.5 build 46348 32bit但是是绿色破解版，放置目录C:\Program Files (x86)\uTorrent3.5.5.46348
都是普通权限执行，跟管理员权限应该关系不大？因为不是完全不能用，现在我就还先用着旧版的
